functional-tests/functional/cpp/tests/AccessorsTest.cpp

```diff
@@ -0,0 +1,195 @@
+    };
+}
+
+} // namespace
+
+TEST( AccessorsOfMutableStruct, GettingFieldsOnLvalue )
+{
+    // Given an object of mutable struct with accessors.
+    SomeMutableStructWithCppAccessors object = get_struct();
+
+    // When using getters on lvalue.
+    const auto& int_field = object.get_trivial_int_field();
+    const auto& double_field = object.get_trivial_double_field();
+    const auto& string_field = object.get_nontrivial_string_field();
+    const auto& list_field = object.get_nontrivial_list_field();
+
+    // Then access can be made.
+    EXPECT_EQ(TrivialInt{32}, int_field);
+    EXPECT_DOUBLE_EQ(TrivialDouble{27.5}, double_field);
+    EXPECT_EQ(NontrivialString{"Some string"}, string_field);
+
+    const NontrivialList expected_list{"S1", "S2", "S3"};
+    EXPECT_EQ(expected_list, list_field);
+}
+
+TEST( AccessorsOfMutableStruct, GettingFieldsOnConstLvalue )
+{
+    // Given const reference to object of mutable struct with accessors.
+    const auto& object = get_const_struct();
+
+    // When using getters on const lvalue.
+    const auto& int_field = object.get_trivial_int_field();
+    const auto& double_field = object.get_trivial_double_field();
+    const auto& string_field = object.get_nontrivial_string_field();
+    const auto& list_field = object.get_nontrivial_list_field();
+
+    // Then access can be made.
+    EXPECT_EQ(TrivialInt{64}, int_field);
+    EXPECT_DOUBLE_EQ(TrivialDouble{52.3}, double_field);
+    EXPECT_EQ(NontrivialString{"Another string"}, string_field);
+
+    const NontrivialList expected_list{"A1", "A2", "A3"};
+    EXPECT_EQ(expected_list, list_field);
+}
+
+TEST( AccessorsOfMutableStruct, GettingFieldsOnRvalue )
+{
+    // When using getter on rvalue.
+    auto int_field = get_struct().get_trivial_int_field();
+    auto double_field = get_struct().get_trivial_double_field();
+    auto string_field = get_struct().get_nontrivial_string_field();
+    auto list_field = get_struct().get_nontrivial_list_field();
+
+    // Then access can be made.
+    EXPECT_EQ(TrivialInt{32}, int_field);
+    EXPECT_DOUBLE_EQ(TrivialDouble{27.5}, double_field);
+    EXPECT_EQ(NontrivialString{"Some string"}, string_field);
+
+    const NontrivialList expected_list{"S1", "S2", "S3"};
+    EXPECT_EQ(expected_list, list_field);
+}
+
+TEST( AccessorsOfMutableStruct, GettingFieldsOnConstRvalue )
+{
+    // When using getters on const rvalue.
+    auto int_field = get_const_struct().get_trivial_int_field();
+    auto double_field = get_const_struct().get_trivial_double_field();
+    auto string_field = get_const_struct().get_nontrivial_string_field();
+    auto list_field = get_const_struct().get_nontrivial_list_field();
+
+    // Then access can be made.
+    EXPECT_EQ(TrivialInt{64}, int_field);
+    EXPECT_DOUBLE_EQ(TrivialDouble{52.3}, double_field);
+    EXPECT_EQ(NontrivialString{"Another string"}, string_field);
+
+    const NontrivialList expected_list{"A1", "A2", "A3"};
+    EXPECT_EQ(expected_list, list_field);
+}
```

Comment:
Note: if these test cases used `auto&&`, then we would have dangling references.